truffle/docs/Safepoints.md

```diff
@@ -0,0 +1,178 @@
+
+* Cancellation, requested exit or interruptions during guest language execution. The stack is unwound by submitting a thread local action.
+* Reading the current stack trace information for other thread than the currently executing thread.
+* Enumerating all object references active on the stack.
+* Implement guest languages that expose a safepoint mechanism as part of their development toolkit.
+* Debuggers evaluating expressions in languages that do not support execution on multiple threads.
+
+## Language Support
+
+Safepoints are explicitly polled by invoking the `TruffleSafepoint.poll(Node)` method. 
+A Truffle guest language implementation must ensure that a safepoint is polled repeatedly within a constant time interval. 
+For example, a single arithmetic expression completes within a constant number of CPU cycles. 
+However, a loop that summarizes values over an array uses a non-constant time dependent on the actual array size. 
+This typically means that safepoints are best polled at the end of loops and at the end of function or method calls to cover recursion. 
+In addition, any guest language code that blocks the execution, like guest language locks, need to use the  `TruffleSafepoint.setBlocked(Interrupter)` API to allow cooperative polling of safepoints while the thread is waiting.
+
+Please read more details what steps languages need to do in the [javadoc](https://www.graalvm.org/truffle/javadoc/com/oracle/truffle/api/TruffleSafepoint.html).
+
+## Thread Local Actions
+
+Languages and instruments can submit action using their environment.
+
+Usage example:
+
+```java
+
+Env env; // language or instrument environment
+
+env.submitThreadLocal(null, new ThreadLocalAction(true /*side-effecting*/, true /*synchronous*/) {
+     @Override
+     protected void perform(Access access) {
+         assert access.getThread() == Thread.currentThread();
+     }
+});
+
+```
+
+Read more in the [javadoc](https://www.graalvm.org/truffle/javadoc/com/oracle/truffle/api/ThreadLocalAction.html).
+
+## Current Limitations
+
+There is currently no way to run thread local actions while the a thread is executing in boundary annotated methods unless the method cooperatively polls safepoints or uses the blocking API. 
+Unfortunately it is not always possible to cooperatively poll safepoints, for example, if the code currently executes third party native code. 
+A future improvement will allow to run code for other threads while they are blocked. 
+This is one of the reasons why it is recommended to use `ThreadLocalAction.Access.getThread()` instead of directly using `Thread.currentThread()`.
+When the native call returns it needs to wait for any thread local action that is currently executing for this thread. 
+This will allow, to collect guest language stack traces from other threads while they are blocked by uncooperative native code.
+Currently the safepoint will be invoked when the native code returns.
```

Comment:
Ruby does some magic with alarm signals and things to interrupt native blocking calls. Is this going to be compatible with that?

Reply:
@eregon  is investigating that. Last time I talked with him, there was no know problems with supporting that.

Reply:
Yes, in Ruby we can interrupt most system calls, notably blocking NFI calls like `read(2)` by sending SIGVTALRM to a given thread (using `pthread_kill()`), and then the thread checks for safepoints before retrying transparently. And C extensions can provide their own Interrupter (Unblocker in TruffleRuby terms), so blocking calls can also be interrupted for C extensions if they use `rb_thread_call_without_gvl()`.